File: qa/jest/__tests__/test_sample.ts
import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';

// Basic tests

describe('Basic tests', () => {
    it('should pass basic test', () => {
        expect(1).toBe(1);
    });
});

// Failing tests

describe('Failing tests', () => {
    it('should throw an error', () => {
        const functionExpectedToThrow = () => {
            throw new Error('Expected error');
        };
        expect(() => functionExpectedToThrow()).toThrowError('Expected error');
    });
});

// Parametrization

describe('Parametrization', () => {
    const cases = [
        [1, 2],
        [2, 3],
    ];

    it.each(cases)('should add 1 to %i to get %i', (input, expected) => {
        expect(input + 1).toBe(expected);
    });
});

// Exceptions

describe('Exceptions', () => {
    it('should raise exception', () => {
        expect(() => {
            throw new Error('ZeroDivisionError');
        }).toThrow('ZeroDivisionError');
    });
});

// Skips

describe('Skips', () => {
    it.skip('should skip this test', () => {
        expect(1).toBe(2);
    });

    const SKIP_IF_TRUE = true;

    if (SKIP_IF_TRUE) {
        it.skip('should skip if true', () => {
            expect(1).toBe(2);
        });
    }
});

// Fixtures

describe('Fixtures', () => {
    let sampleFixture: number;

    beforeEach(() => {
        sampleFixture = 42;
    });

    it('should use fixture', () => {
        expect(sampleFixture).toBe(42);
    });

    let sampleFixtureWithTeardown: number;

    beforeEach(() => {
        sampleFixtureWithTeardown = 42;
    });

    afterEach(() => {
        console.log('teardown');
    });

    it('should use fixture with teardown', () => {
        expect(sampleFixtureWithTeardown).toBe(42);
    });
});

File: qa/pytest/tests/test_sample.py
import pytest

# Basic tests

def test_basic():
    assert 1 == 1

# Failing tests

@pytest.mark.xfail(reason="deliberate fail")
def test_fail():
    assert 1 == 2

# Parametrization

@pytest.mark.parametrize("input,expected", [
    (1, 2),
    (2, 3),
])
def test_param(input, expected):
    assert input + 1 == expected

# Exceptions

def test_raise_exception():
    with pytest.raises(ZeroDivisionError):
        1 / 0

# Skips

@pytest.mark.skip(reason="deliberate skip")
def test_skip():
    assert 1 == 2

SKIP_IF_TRUE = True

@pytest.mark.skipif(SKIP_IF_TRUE, reason="skip if true")
def test_skip_if_true():
    assert 1 == 2

# Fixtures

@pytest.fixture()
def sample_fixture():
    return 42

def test_fixture(sample_fixture):
    assert sample_fixture == 42

@pytest.fixture()
def sample_fixture_with_teardown():
    yield 42
    print("teardown")

def test_fixture_with_teardown(sample_fixture_with_teardown):
    assert sample_fixture_with_teardown == 42

# Markers
# To run tests with a specific marker, use the -m option: pytest -m foo_marker

@pytest.mark.foo_marker
def test_marker():
    assert 1 == 1

# combine multiple marks using logical operators: pytest -m "foo_marker or bar_marker"

@pytest.mark.bar_marker
def test_marker2():
    assert 1 == 1
